compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -213,6 +220,95 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         bug!();
     }
 
+    fn maybe_report_recursive_gat(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        parent: &PredicateObligation<'tcx>,
+        child: &PredicateObligation<'tcx>,
+    ) -> bool {
+        let (parent, child) =
+            match (parent.predicate.kind().skip_binder(), child.predicate.kind().skip_binder()) {
+                (ty::PredicateKind::Trait(head, ..), ty::PredicateKind::Trait(tail, ..)) => {
+                    (head, tail)
+                }
+                _ => return false,
+            };
+
+        if Some(parent.clone().def_id()) != self.tcx.lang_items().sized_trait() {
+            return false;
+        }
+
+        let parent_ty = parent.self_ty();
+        let child_ty = child.self_ty();
+        debug!(
+            "report_overflow_error_cycle: parent_ty = {:?}, child_ty = {:?}",
+            parent_ty, child_ty
+        );
+
+        let parent_kind = parent_ty.kind();
+        let child_kind = child_ty.kind();
+        debug!(
+            "report_overflow_error_cycle: parent_kind = {:?}, child_kind = {:?}",
+            parent_kind, child_kind
+        );
+
+        let (substs, item_def_id) = match parent_kind {
+            ty::Projection(ty::ProjectionTy { substs, item_def_id }) => (substs, item_def_id),
+            _ => return false,
+        };
+
+        let hir_id = self.tcx.hir().local_def_id_to_hir_id(item_def_id.expect_local());
+        let node = self.tcx.hir().get(hir_id);
+        debug!("report_overflow_error_cycle: node = {:?}", node);
+
+        let (assoc_generics, _assoc_span) = match node {
+            Node::TraitItem(hir::TraitItem {
+                generics,
+                kind: hir::TraitItemKind::Type(..),
+                span,
+                ..
+            }) => (generics, span),
+            _ => return false,
+        };
+
+        let parameter_size = assoc_generics.params.len();
+        if parameter_size == 0 {
+            return false;
+        }
+
+        assert!(substs.len() >= parameter_size);
+
+        let index = match substs
+            .get(substs.len() - parameter_size..)
+            .map(|gens| {
+                gens.iter()
+                    .enumerate()
+                    .find_map(|(i, param)| if param == &child_ty.into() { Some(i) } else { None })
+            })
+            .flatten()
+        {
+            Some(index) => index,
+            None => return false,
+        };
+        debug!("report_overflow_error_cycle: index = {:?}", index);
+
+        let param = &assoc_generics.params[index];
```

Comment:
This all feels a bit weird to me. But running out of time right now to think more about this.